### PR TITLE
loader: Duplicate file descriptors instead of re-opening by filename in archives

### PR DESCRIPTION
**Pull request by uyjulian**
This avoids additional unnecessary filesystem lookups when opening files in an archive.

### CODE REVIEW

**Comment by renpytom** [comment]
Do you have any benchmarks on this? Opening a cached file should be fast, and youy're making a trip to kernel space either way.

**Comment by uyjulian** [comment]
Here is a benchmark:
```
julian@julians-MacBook-Pro:/Volumes/public_share/tst$ python3 -m timeit --number=100 "with open('hoge.txt', 'rb') as f: pass"
100 loops, best of 5: 1.68 msec per loop
julian@julians-MacBook-Pro:/Volumes/public_share/tst$ python3 -m timeit --number=100 --setup="import os"$'\n'"f = open('hoge.txt')" "open(os.dup(f.fileno()), 'rb')"
100 loops, best of 5: 15.6 usec per loop
julian@julians-MacBook-Pro:/Volumes/public_share/tst$ 
```

**Review by mal — COMMENTED** [review]
Out of curiosity I tried to replicate this on an Ubuntu machine but couldn't get close to the same results. Might be a platform thing, or me doing something wrong, but running slightly modified benches on Ubuntu yielded identical timings in the two approaches.

```
$ python3 -m timeit "with open('junk.bin', 'rb') as f: f.seek(0)" 
100000 loops, best of 3: 10.6 usec per loop

$ python3 -m timeit -s "import os; fd = open('junk.bin')" "f = open(os.dup(fd.fileno()), 'rb'); f.seek(0)"
100000 loops, best of 3: 10.6 usec per loop
```

In case it matters `junk.bin` was a ~310M file of random noise.

**Comment by renpytom** [comment]
That's a convincing improvement. 

My only question is about what you were thinking with the various:

```
except OSError:
     pass
```

clauses? Wouldn't it make sense to show the error? Or am I missing something?

File: renpy/loader.py [hunk 1]
```diff
@@ -610,13 +623,18 @@ def write(self, s):
...
+            f.seek(0, 2)
+            length = f.tell()
+            f.seek(0, 0)
+            try:
+                fd = os.dup(f.fileno())
+            except OSError:
+                pass
 
-        return SubFile(f, 0, length, b'')
+        return SubFile(fn=name, base=0, length=length, start=b'', fd=fd)
```
[the remark this answers is not in the document]
Comment (uyjulian) [reply renpy/loader.py:637]: The order of arguments was wrong here, so I decided to add kwargs here. I think the clarity improvement with kwargs here is OK.

**Comment by uyjulian** [comment]
> That's a convincing improvement.
> 
> My only question is about what you were thinking with the various:
> 
> ```
> except OSError:
>      pass
> ```
> 
> clauses? Wouldn't it make sense to show the error? Or am I missing something?

The `except OSError` clauses are for the cases where `.fileno()` does not work (e.g. the underlying stream does not use a file descriptor). In that case it would allow falling back to the old behavior of file name lookup.



> Out of curiosity I tried to replicate this on an Ubuntu machine but couldn't get close to the same results. Might be a platform thing, or me doing something wrong, but running slightly modified benches on Ubuntu yielded identical timings in the two approaches.
> 
> ```
> $ python3 -m timeit "with open('junk.bin', 'rb') as f: f.seek(0)" 
> 100000 loops, best of 3: 10.6 usec per loop
> 
> $ python3 -m timeit -s "import os; fd = open('junk.bin')" "f = open(os.dup(fd.fileno()), 'rb'); f.seek(0)"
> 100000 loops, best of 3: 10.6 usec per loop
> ```
> 
> In case it matters `junk.bin` was a ~310M file of random noise.

This was on a NFS share on macOS 10.14.6.

**Comment by renpytom** [comment]
What are the numbers you're getting on a normal disk, rather than an NFS share? I don't think we should make Ren'Py more complex for an obscure case like this. 

On my computers: 

Linux: 2.6 us (open) vs 2.51 us (fileno)
Mac: 35 us vs 16 us
Windows: 62.8 usec vs 20.5 usec
Window (Samba-mounted share from Linux): 447 usec vs 225 usec. 

With these numbers, I'm a lot less convinced. This is especially true since file operations don't really take place in a performance critical environment. 

(I'd also think that if we decide to address this, it probably makes sense to not open the files at all. The real performance-criticial stuff is in pygame_sdl2's rwobject.)

**Comment by renpytom** [comment]
This will probably be better served by a fix to #3342. So closing this.

**Pull request closed without merging**